src/ray/raylet/dependency_manager.cc

```diff
@@ -0,0 +1,309 @@
+  return local_objects_.count(object_id) == 1;
+}
+
+bool DependencyManager::GetOwnerAddress(const ObjectID &object_id,
+                                        rpc::Address *owner_address) const {
+  auto obj = required_objects_.find(object_id);
+  if (obj == required_objects_.end()) {
+    return false;
+  }
+
+  *owner_address = obj->second.owner_address;
+  return !owner_address->worker_id().empty();
+}
+
+void DependencyManager::RemoveObjectIfNotNeeded(
+    absl::flat_hash_map<ObjectID, DependencyManager::ObjectDependencies>::iterator
+        required_object_it) {
+  const auto &object_id = required_object_it->first;
+  if (required_object_it->second.Empty()) {
+    RAY_LOG(DEBUG) << "Object " << object_id << " no longer needed";
+    if (required_object_it->second.wait_request_id > 0) {
+      RAY_LOG(DEBUG) << "Canceling pull for wait request of object " << object_id
+                     << " request: " << required_object_it->second.wait_request_id;
+      object_manager_.CancelPull(required_object_it->second.wait_request_id);
+    }
+    if (!local_objects_.count(object_id)) {
+      reconstruction_policy_.Cancel(object_id);
+    }
+    required_objects_.erase(required_object_it);
+  }
+}
+
+absl::flat_hash_map<ObjectID, DependencyManager::ObjectDependencies>::iterator
+DependencyManager::GetOrInsertRequiredObject(const ObjectID &object_id,
+                                             const rpc::ObjectReference &ref) {
+  auto it = required_objects_.find(object_id);
+  if (it == required_objects_.end()) {
+    it = required_objects_.emplace(object_id, ref).first;
+    if (local_objects_.count(object_id) == 0) {
+      reconstruction_policy_.ListenAndMaybeReconstruct(object_id, ref.owner_address());
+    }
+  }
+  return it;
+}
+
+void DependencyManager::StartOrUpdateWaitRequest(
+    const WorkerID &worker_id,
+    const std::vector<rpc::ObjectReference> &required_objects) {
+  RAY_LOG(DEBUG) << "Starting wait request for worker " << worker_id;
+  auto &wait_request = wait_requests_[worker_id];
+  for (const auto &ref : required_objects) {
+    const auto obj_id = ObjectRefToId(ref);
+    if (local_objects_.count(obj_id)) {
+      // Object is already local. No need to fetch it.
+      continue;
+    }
+
+    if (wait_request.insert(obj_id).second) {
+      RAY_LOG(DEBUG) << "Worker " << worker_id << " called ray.wait on non-local object "
+                     << obj_id;
+      auto it = GetOrInsertRequiredObject(obj_id, ref);
+      it->second.dependent_wait_requests.insert(worker_id);
+      if (it->second.wait_request_id == 0) {
+        it->second.wait_request_id = object_manager_.Pull({ref});
+        RAY_LOG(DEBUG) << "Started pull for wait request for object " << obj_id
+                       << " request: " << it->second.wait_request_id;
+      }
+    }
+  }
+
+  // No new objects to wait on. Delete the empty entry that was created.
+  if (wait_request.empty()) {
+    wait_requests_.erase(worker_id);
```

Comment:
When do we need this? Aren't we supposed to always explicitly call CancelWaitRequest to remove wait requests?

Reply:
We only keep wait requests open until the object becomes local, so a worker is not required to explicitly cancel a previous wait request. That's why we erase here, to prevent a leak of the worker entry (actually I missed this before and it was caught by one of my unit tests :) ).